src/client/particles.cpp

```diff
@@ -31,635 +29,808 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	virtual void setMaxStartColor(const video::SColor& /* color */) {}
+	virtual void setMaxLifeTime(const u32 /* t */) {}
+	virtual void setMinLifeTime(const u32 /* t */) {}
+	virtual u32 getMaxLifeTime() const {return 1;}
+	virtual u32 getMinLifeTime() const {return 1;}
+	virtual void setMaxAngleDegrees(const s32 /* t */) {}
+	virtual s32 getMaxAngleDegrees() const {return 0;}
+	virtual void setMaxStartSize(const core::dimension2df& /* size */) {}
+	virtual void setMinStartSize(const core::dimension2df& /* size */) {}
+	virtual void setCenter(const core::vector3df& /* center */) {}
+	virtual void setRadius(f32 /* radius */) {}
+	virtual const core::vector3df& getDirection() const {return direction;}
+	virtual u32 getMinParticlesPerSecond() const {return 1;}
+	virtual u32 getMaxParticlesPerSecond() const {return 1;}
+	virtual const video::SColor& getMinStartColor() const {return minStartColor;}
+	virtual const video::SColor& getMaxStartColor() const {return maxStartColor;}
+	virtual const core::dimension2df& getMaxStartSize() const {return max_size;}
+	virtual const core::dimension2df& getMinStartSize() const {return min_size;}
+	virtual const core::vector3df& getCenter() const {return center;}
+	virtual f32 getRadius() const {return 1.f;}
+	virtual irr::scene::E_PARTICLE_EMITTER_TYPE getType() const
+	{
+		return irr::scene::EPET_COUNT;
+	}
+protected:
+	core::dimension2df min_size, max_size;
+private:
+	// unused
+	const v3f direction;
+	const video::SColor minStartColor, maxStartColor;
+	const v3f center;
+};
+
+class SingleEmitter : public CustomEmitter
 {
-	v3f ppos = m_player->getPosition() / BS;
-	v3f pos = random_v3f(m_minpos, m_maxpos);
-
-	// Need to apply this first or the following check
-	// will be wrong for attached spawners
-	if (is_attached) {
-		pos.rotateXZBy(attached_yaw);
-		pos += attached_pos;
+public:
+	SingleEmitter(irr::scene::ISceneManager *smgr,
+		irr::scene::IParticleSystemSceneNode *ps, const video::SColor &color, u32 number) :
+		smgr(smgr), ps(ps), number(number), deletion_time(0), color(color),
+		random_properties(true)
+	{
+		particles.reserve(number);
+		++single_particles_count;
 	}
 
-	if (pos.getDistanceFrom(ppos) > radius)
-		return;
+	SingleEmitter(irr::scene::ISceneManager *smgr,
+		irr::scene::IParticleSystemSceneNode *ps, const video::SColor &color,
+		f32 expirationtime, f32 size, const v3f &vel):
+		smgr(smgr), ps(ps), number(1), deletion_time(0), color(color),
+		expirationtime((u32) (expirationtime * 1e3f)), size(size), pos(0), vel(vel),
+		random_properties(false)
+	{
+		++single_particles_count;
+	}
 
-	v3f vel = random_v3f(m_minvel, m_maxvel);
-	v3f acc = random_v3f(m_minacc, m_maxacc);
+	~SingleEmitter()
+	{
+		--single_particles_count;
+	}
 
-	if (is_attached) {
-		// Apply attachment yaw
-		vel.rotateXZBy(attached_yaw);
-		acc.rotateXZBy(attached_yaw);
+	virtual s32 emitt(u32 now, u32 timeSinceLastCall, irr::scene::SParticle *&outArray)
```

Comment:
Why the two `t`s?

Reply:
That's a misspelling in the Irrlicht API...
http://irrlicht.sourceforge.net/docu/classirr_1_1scene_1_1_i_particle_emitter.html